Feature: generated injector delegator

michalbundyra opened the pull request:
In documentation we are suggesting to create in manually. I think we can have it in the library as it's generic code.

See: https://docs.zendframework.com/zend-di/cookbook/aot-guide/#5-add-aot-to-the-service-manager

Documentations needs update if it is going to be accepted.

- [X] Are you creating a new feature?
  - [X] Why is the new feature needed? What purpose does it serve?
  - [X] How will users use the new feature?
  - [X] Base your feature on the `develop` branch, and submit against that branch.
  - [X] Add only one feature per pull request; split multiple features over multiple pull requests
  - [X] Add tests for the new feature.
  - [x] Add documentation for the new feature.
  - [x] Add a `CHANGELOG.md` entry for the new feature.

michalbundyra committed:
Added GeneratedInjectorDelegator

As it could be used comonnly would be nice to have it out of the box.

tux-rampage commented on src/GeneratedInjectorDelegator.php:
```diff
@@ -0,0 +1,38 @@
+use Interop\Container\ContainerInterface;
+use Zend\ServiceManager\Factory\DelegatorFactoryInterface;
+
+class GeneratedInjectorDelegator implements DelegatorFactoryInterface
```
Tend service manager is no dependency of zend-di, therefore factories should not rely on its interfaces to be present. This should be a generic invokable accepted by zend service manager.

```suggestion
class GeneratedInjectorDelegator
```

tux-rampage commented on src/GeneratedInjectorDelegator.php:
```diff
@@ -0,0 +1,38 @@
+        ContainerInterface $container,
+        $name,
+        callable $callback,
+        array $options = null
```
`$options` is not used and can be removed.

tux-rampage reviewed (changes requested):
This makes sense. Thanks a lot for your effort. Just a view nitpicks that needs to be fixed before a merge.

I also think it makes sense to add it to the `ConfigProvider`

michalbundyra committed:
Drop interface and useless parameter in delegator

- per @tux-rampage

webimpress commented:
@tux-rampage 

> I also think it makes sense to add it to the `ConfigProvider`

I am not sure about it, what if someone already has it added there (the custom one) or if someone what to use another?

tux-rampage commented:
> @tux-rampage
> 
> > I also think it makes sense to add it to the `ConfigProvider`
> 
> I am not sure about it, what if someone already has it added there (the custom one) or if someone what to use another?

Good point. Could you update the AoT guide?

michalbundyra committed:
Updated docs to use new decorator

webimpress commented:
@tux-rampage 

> Could you update the AoT guide?

Docs is updated now.

tux-rampage commented on src/GeneratedInjectorDelegator.php:
```diff
@@ -0,0 +1,32 @@
+
+        $injector = $callback();
+
+        $generatedInjector = $namespace . '\\GeneratedInjector';
```
I think we should somehow handle a null or non-string namespace value. by extracting this to a method that ensures typesafety:

```php
private function buildDecoratorClassName(string $namespace): string
{
    return $namespace !== '' ? sprintf('%s\\GeneratedInjector', $namespace) : 'GeneratedInjector';
}
```

and then:

```suggestion
        if ($namespace === null) {
            return $injector;
        }
        
        $generatedInjector = $this->buildDecoratorClassName($namespace);
```

webimpress replied:
No needed, if namespace is empty then `GeneratedInjector` is in the global namespace so `\GeneratedInjector` is perfectly fine.

tux-rampage commented on src/GeneratedInjectorDelegator.php:
```diff
@@ -0,0 +1,32 @@
+    {
+        $config = $container->has('config') ? $container->get('config') : [];
+        $aotConfig = $config['dependencies']['auto']['aot'] ?? [];
+        $namespace = $aotConfig['namespace'] ?? null;
```
`Zend\Di\CodeGenerator\InjectorGenerator` defaults the namespace to `'Zend\Di\Generated'` on [line 98](https://github.com/zendframework/zend-di/blob/develop/src/CodeGenerator/InjectorGenerator.php#L98), I think this should do the same here as well.

webimpress replied:
Also disagree here, this is the namespace from the configuration, we are using later on. The only think we can change is to set it to empty string. It's not making any difference here: as we using it later in the concatenation.

webimpress replied:
We have done here the same what is in the other place:

https://github.com/zendframework/zend-di/blob/7a686b016719069862d225c3dfb6d170646c99c8/src/Container/GeneratorFactory.php#L29-L33

tux-rampage reviewed (changes requested):
I found some additional minor things that we may need to address.

Thanks for your effort to improve this component :+1:

michalbundyra committed:
If namespace is not configured/empty use default one: Zend\Di\Generated

This namespace is default for generated resources.

michalbundyra committed:
Added import internal PHP function: class_exists

michalbundyra committed:
Check if provided namespace is a string

webimpress commented on src/GeneratedInjectorDelegator.php:
```diff
@@ -0,0 +1,42 @@
+
+        if (! is_string($namespace)) {
+            throw new class('Provided namespace is not a string.') extends UnexpectedValueException
+                implements ContainerExceptionInterface
```
Not sure what exception should be thrown here. I think it should be instance of `ContainerExceptionInterface` as delegator is used in PSR-11 container.

tux-rampage replied:
`ContainerExceptionInterface` is fine, but please make it a named class in the `Zend\Di\Exception` namespace. This will also solve the PHPUnit issue.

webimpress commented on test/GeneratedInjectorDelegatorTest.php:
```diff
@@ -0,0 +1,73 @@
+
+        $delegator = new GeneratedInjectorDelegator();
+
+        $this->expectException(ContainerExceptionInterface::class);
```
Testing here just for `ContainerExceptionInterface`, using PHPUnit methods we can't check two types, we can do it using custom `try...catch`, but not sure if it is needed ?

tux-rampage replied:
If you make this exception a named class, as requested above, you can check for it here and this wil no longer be an issue.

tux-rampage reviewed (changes requested):
Hi @webimpress 

sorry for the delay on this. Just one thing with the exception you throw. After that I think we're good to go.

tux-rampage committed:
Change anonymous to named exception class

This will improve readability and makes the unit easier to test.
It will also offer a typed information to consumers to get rid
it.

tux-rampage committed:
Remove unused class imports

tux-rampage commented:
@michalbundyra As discussed in slack, I've addressed this one. Can you pleas re-check?

michalbundyra commented on src/GeneratedInjectorDelegator.php:
```diff
@@ -0,0 +1,38 @@
+    {
+        $config = $container->has('config') ? $container->get('config') : [];
+        $aotConfig = $config['dependencies']['auto']['aot'] ?? [];
+        $namespace = empty($aotConfig['namespace']) ? 'Zend\Di\Generated' : $aotConfig['namespace'];
```
Maybe better "empty or non-string" here?

tux-rampage replied:
```suggestion
        $namespace = empty($aotConfig['namespace']) || !is_string($aotConfig['namespace']) 
            ? 'Zend\Di\Generated' 
            : $aotConfig['namespace'];
```

Like this?

michalbundyra replied:
yes, please

test would be with an array, for example, as it can't be casted to string

tux-rampage replied:
I was just going to make this change and saw it would render the next 3 lines useless:

```php
        $config = $container->has('config') ? $container->get('config') : [];
        $aotConfig = $config['dependencies']['auto']['aot'] ?? [];
        $namespace = empty($aotConfig['namespace']) ? 'Zend\Di\Generated' : $aotConfig['namespace'];

        if (! is_string($namespace)) {
            throw new InvalidServiceConfigException('Provided namespace is not a string.');
        }
```

I'd argue to leave it as is. This will be explicit to the user that there is an unusable misconfiguration instead of hiding it behind a default value that may cause failures which may become hard to debug. At least it's not obvious what's the issue that cause incorrect behavior.

I'll change the test from an int fixture to an array though.

michalbundyra commented on src/Exception/InvalidServiceConfigException.php:
```diff
@@ -0,0 +1,14 @@
+
+use Psr\Container\ContainerExceptionInterface;
+
+class InvalidServiceConfigException extends LogicException implements ContainerExceptionInterface
```
not sure if LogicException as a parent is the correct one

Maybe better InvalidArgumentException?

tux-rampage replied:
InvalidArgumentException actually is a LogicException.
But in this case I'd argue that this is not cased by a function argument but a (service)configuration value. This could also be seen a `RuntimeException`.

I chose LogicException since it's the developer's job to configure the services correctly, and if they don't provide a namespace-string, they made a logical mistake.

> https://www.php.net/manual/en/class.logicexception.php
> Exception that represents error in the program logic. This kind of exception should lead directly to a fix in your code. 

But I'm happy if you provide a better fit.

michalbundyra replied:
Ok, sounds good to me.

I just wonder if we should have then another exception before if the namespace is provided but is not null and not a string? Maybe we shouldn't ignore invalid value. What do you think?

tux-rampage replied:
Isn't this already the case or am I missing something? :thinking:

michalbundyra commented on test/GeneratedInjectorDelegatorTest.php:
```diff
@@ -27,7 +28,7 @@ public function testProvidedNamespaceIsNotAString()
         $delegator = new GeneratedInjectorDelegator();
 
-        $this->expectException(ContainerExceptionInterface::class);
+        $this->expectException(InvalidServiceConfigException::class);
```
it was quite important check, as container (per specification) can throw only `ContainerExceptionInterface` instances.

would be nice to have it somewhere

tux-rampage replied:
Is a unit test really the correct scope? We could assert that `InvalidServiceConfigException` implements `ContainerExceptionInterface`. So that whenever this changes this test fails:

```suggestion
        self::assertInstanceOf(ContainerInterface::class, new InvalidServiceConfigException());
        $this->expectException(InvalidServiceConfigException::class);
```

michalbundyra replied:
Yeah, probably it should be separate test case, but I am happy to have it here for time being.

tux-rampage replied:
I guess a separate test would be out of scope for now. I'm not entirely sure if this is the correct place or way to cover this. Maybe we find something more appropriate in the future. :wink: 

Until then I added the assert as suggested.

michalbundyra commented on docs/book/cookbook/aot-guide.md:
````diff
@@ -297,36 +297,10 @@ the following files:
-```
-
-After this we need to add configuration to the `ConfigProvider` class we created
+First, we'll use a delegator `Zend\Di\GeneratedInjectorDelegator` to decorate
````
I think we need to note that `GeneratedInjectorDelegator` is available since version X.Y.Z.

tux-rampage replied:
Absolutely

tux-rampage replied:
Docs are updated. If somebody has a more elegant phrasing or suggesstion, I'd appreciate to hear it.

tux-rampage commented:
Static analysis is failing: https://travis-ci.org/zendframework/zend-di/jobs/625613032#L425
This was introduced in #56

tux-rampage committed:
Add changelog entry

tux-rampage committed:
Add information for the delegator factory before 3.2

tux-rampage committed:
Add test to ensure InvalidServiceConfigException implements container exception

tux-rampage committed:
Change test fixture from int to array

michalbundyra committed:
Merge branch 'develop' into feature/generated-injector-delegator
